### ISSUE TITLE
should `EitherAsync` `.join` and `.ap` widen the left-side type?

### ISSUE THREAD

**Issue by atomanyih** [gigobyte/purify#526]
Current definition:
```ts
interface EitherAsync<L, R> {
  ap<R2>(other: EitherAsync<L, (value: R) => R2>): EitherAsync<L, R2>;
  join<R2>(this: EitherAsync<L, EitherAsync<L, R2>>): EitherAsync<L, R2>;
}
```

Proposed definition:
```ts
interface EitherAsync<L, R> {
  ap<L2, R2>(other: EitherAsync<L2, (value: R) => R2>): EitherAsync<L | L2, R2>;
  join<L2, R2>(this: EitherAsync<L, EitherAsync<L2, R2>>): EitherAsync<L | L2, R2>;
}
```

## Why `.ap`?
`.ap` doesn't work if the function being applied is an `EitherAsync` that has a different, nominally-typed left side.

```ts
type NoMoreHotdogsError = {
  type: 'NoMoreHotdogs';
  message: string;
};

type UserNotHomeError = {
  type: 'UserNotHomeError';
  message: string;
};

type UserNotHungryError = {
  type: 'UserNotHungryError';
  message: string;
};

function getHotdog(): EitherAsync<NoMoreHotdogsError, Hotdog> 

function getUser(): EitherAsync<NotHomeError, User>

function feedUserHotdog(
  hotdog: Hotdog,
  user: User
): EitherAsync<UserNotHungryError, User>

getUser()  
  .ap(  
    getHotdog()  
      .map(feedUserHotdog)  
  );
/*
TS2345: Argument of type 'EitherAsync<NoMoreHotdogsError, (user: User) => EitherAsync<UserNotHungryError, User>>' is not assignable to parameter of type 'PromiseLike<Either<NotHomeError, (value: User) => EitherAsync<UserNotHungryError, User>>>'.
  ** etc **
    Type 'NoMoreHotdogsError' is not assignable to type 'NotHomeError'. 
*/
```

If in the above example I opted to use error types that were structurally equivalent, it compiles but then `.ap` throws away the inner left side type.

```ts
type NoMoreHotdogsError = {
  message: string;
};

type UserNotHomeError = {
  message: string;
};

type UserNotHungryError = {
  message: string;
};

function getHotdog(): EitherAsync<NoMoreHotdogsError, Hotdog> 

function getUser(): EitherAsync<NotHomeError, User>

function feedUserHotdog(
  hotdog: Hotdog,
  user: User
): EitherAsync<UserNotHungryError, User>

getUser()  
  .ap(  
    getHotdog()  
      .map(feedUserHotdog)  
// Typed as:
// EitherAsync<UserNotHomeError, EitherAsync<UserNotHungryError, User>>

// More useful to be:
// EitherAsync<UserNotHomeError | NoMoreHotdogsError, EitherAsync<UserNotHungryError, User>>
  );
```

## Why `.join`?
Continuing on the above example, if you `.join` to flatten the nested `EitherAsync`,  it discards the left-side information of the inner.

```ts
getUser()  
  .ap(  
    getHotdog()  
      .map(feedUserHotdog)  
// EitherAsync<UserNotHomeError, EitherAsync<UserNotHungryError, User>>
  ).join();
// EitherAsync<UserNotHomeError, User>
```

You should theoretically be able to use `.chain(a => a)` here to unwrap and preserve the inner left type

```ts
getUser()  
  .ap(  
    getHotdog()  
      .map(feedUserHotdog)  
// EitherAsync<UserNotHomeError, EitherAsync<UserNotHungryError, User>>
  ).chain(x => x);
// EitherAsync<UserNotHomeError | UserNotHungryError, User>
```

However, `.chain` [causes problems with nested `EitherAsync`](https://github.com/gigobyte/purify/issues/525).

So, changing the type of `.ap` and `.join` would both allow you to use nominally-typed left side types, and reflect more truthfully why the `EitherAsync` failed. That is, the reason the user did not eat the hot dog could be one of: they weren't home, there were no hotdogs, or they weren't hungry.

It also means that `.join` and `.chain(x => x)` would be equivalently typed.

------
For the last week I've been using this library and evaluating it for use in our prod app and have really appreciated its approachability and clear code examples. It has been much easier to learn than the brick wall of fp-ts!

However, I ran into a few problems that have prevented us from easily adopting it in our production code

**Comment by gigobyte** [comment]
Yep, this follows the same logic as with `chain` and the other methods that create an union out of the errors so I agree it should be implemented. Thanks for your contribution :)

**Comment by gigobyte** [comment]
Implemented in `v1.2.0`.
